torchtune/datasets/_finetune.py

```diff
@@ -0,0 +1,146 @@
+    Hugging Face Hub, local files, or remote files. This class supports instruct,
+    chat, tool, or multimodal data for fine-tuning. At a high level, this class
+    will load the data from source and apply the following pre-processing steps
+    when a sample is retrieved:
+
+    1. Dataset-specific transform. This is typically unique to each dataset and extracts
+       the necessary columns into torchtune's :class:`~torchtune.data.Message` format,
+       a standardized API for all model tokenizers.
+    2. If specified, apply a prompt template for the task you are fine-tuning for.
+    3. Model-specific transform or tokenization
+
+    All datasets are formatted into :class:`~torchtune.data.Message`s because for
+    fine-tuning, datasets can be considered as "conversations" with the model,
+    or AI assistant. Thus, we can standardize all text content as messages in a conversation assigned to
+    a :class:`~torchtune.data.Role`:
+
+    - system messages contain the system prompt
+    - user messages contain the input prompt into the model
+    - assistant messages are the response of the model and what you actually want
+      to train for and compute loss directly against
+    - ipython messages are the return from a tool call
+
+    Chat datasets are multiple rounds of user-assistant messages. Instruct datasets
+    are typically a single round involving a specific instruction and the model's response.
+    Tool datasets are a type of chat dataset that includes ipython messages. Multimodal
+    datasets are a type of chat dataset that incorporates media into the user messages.
+
+    The :class:`~torchtune.data.Message` forms the core data unit that all tokenizer
+    APIs expect. The key component of this class that ensures any dataset is transformed
+    into this format is the ``message_transform``. This is a callable class that takes
+    in a sample dictionary - typically a single row from the source dataset - that
+    processes the sample in any configurable way to output a list of messages::
+
+        [
+            Message(
+                role=<system|user|assistant|ipython>,
+                content=<message>,
+            ),
+            ...
+        ]
+
+    For any custom dataset, use the ``message_transform`` to contain all pre-processing to
+    return the list of messages.
+
+    Any model-specific pre-processing that needs to happen can be configured with the ``model_transform``
+    parameter. This is another callable class that contains any custom logic tied to the
+    model you are fine-tuning and will carry over to inference. For example, text + image
+    multimodal datasets requires processing the images in a way specific to the vision
+    encoder being used by the model and is agnostic to the specific dataset.
+
+    Tokenization is handled by the ``model_transform``. All :class:`~torchtune.modules.tokenizers.ModelTokenizer`s
```

Comment:
:class:`~torchtune.modules.tokenizers.ModelTokenizer` is not rendering correctly :(

Reply:
because we don't actually generate the doc for that anywhere perhaps?